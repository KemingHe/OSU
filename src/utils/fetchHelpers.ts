// ./src/utils/fetchHelpers.ts
//
// Helper functions for fetching and querying data.

// Local constant import.
import UserAgents from "@src/constants/userAgents";

// -----------------------------------------------------------------------------
export function getTextContentFromElement(
  element: Element,
): string | undefined {
  // Short-circuit to return undefined if the element has no text content.
  const textContent: string | null = element.textContent;

  // Once element is defined (not null), textContent will never be null,
  // but must be checked for typescript. Thus, ignore next 3 unreachable lines.
  /* v8 ignore next 3 */
  if (textContent === null) {
    return undefined;
  }

  // Short-circuit to return undefined if the text content is empty after sanitization.
  const sanitizedTextContent: string = sanitizeText(textContent);
  if (sanitizedTextContent === "") {
    return undefined;
  }

  return sanitizedTextContent;
}

// -----------------------------------------------------------------------------
export interface GetTextContentFromSelectorOptions {
  element: Element;
  selector: string;
}

export function getTextContentFromSelector({
  element,
  selector,
}: GetTextContentFromSelectorOptions): string | undefined {
  // Short-circuit to return undefined if the target element is not found.
  const targetElement: Element | null = element.querySelector(selector);
  if (targetElement === null) {
    return undefined;
  }

  return getTextContentFromElement(targetElement);
}

// -----------------------------------------------------------------------------
export interface GetAttributeFromSelectorOptions {
  element: Element;
  selector: string;
  attribute: string;
}

export function getAttributeFromSelector({
  element,
  selector,
  attribute,
}: GetAttributeFromSelectorOptions): string | string | undefined {
  // Short-circuit to return undefined if the target element is not found.
  const targetElement: Element | null = element.querySelector(selector);
  if (targetElement === null) {
    return undefined;
  }

  // Short-circuit to return undefined if the target attribute is not found.
  const targetAttribute: string | null = targetElement.getAttribute(attribute);
  if (targetAttribute === null) {
    return undefined;
  }

  // Short-circuit to return undefined if the attribute is empty after sanitization
  const sanitizedAttribute: string = sanitizeText(targetAttribute);
  if (sanitizedAttribute === "") {
    return undefined;
  }

  return sanitizedAttribute;
}

// -----------------------------------------------------------------------------
export function sanitizeText(text: string): string {
  // Remove double quotes around the entire text.
  const qoutesRemovedText: string = text.trim().replace(/^"(.*)"$/, "$1");

  // Remove commas and periods at the start and end of the text.
  const commaDotRemovedText: string = qoutesRemovedText
    .trim()
    .replace(/^[,\.]+|[,\.]+$/g, "");

  // Remove last remaining whitespaces at the start and end of the text.
  return commaDotRemovedText.trim();
}

// -----------------------------------------------------------------------------
export interface RandomDelayOptions {
  min: number;
  max: number;
}

export async function randomDelay({
  min,
  max,
}: RandomDelayOptions): Promise<void> {
  // Handle invalid input by throwing an error.
  if (min < 0) {
    throw new Error("Minimum delay must be greater than or equal to 0.");
  }

  if (max <= 0) {
    throw new Error("Maximum delay must be strictly greater than 0.");
  }

  if (max < min) {
    throw new Error(
      "Maximum delay must be greater than or equal to minimum delay.",
    );
  }

  // ---------------------------------------------------------------------------
  // Short-circuit to return immediately if min and max are the same.
  if (min === max) {
    return new Promise((resolve) => setTimeout(resolve, min));
  }

  // ---------------------------------------------------------------------------
  // Generate a random delay between min and max milliseconds.
  const delay: number = Math.floor(Math.random() * (max - min + 1) + min);
  return new Promise((resolve) => setTimeout(resolve, delay));
}

// -----------------------------------------------------------------------------
export function randomUserAgent(): string {
  const randomAgent: string | undefined =
    UserAgents[Math.floor(Math.random() * UserAgents.length)];

  // randomeAgent is always defined, but must be checked for typescript.
  /* v8 ignore next 3 */
  if (!randomAgent) {
    throw new Error("Failed to generate random user agent.");
  }
  return randomAgent;
}
