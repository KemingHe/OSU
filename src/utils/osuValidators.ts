// ./src/utils/osuValidators.ts
//
// OSU-related validators using regex.

/**
 * Regular expression for validating OSU name.number format.
 * @type {RegExp}
 */
const osuNameDotNumberRegExp: RegExp = /^[a-z]+\.[1-9]\d*$/;

/**
 * Regular expression for validating OSU email addresses.
 * @type {RegExp}
 */
const osuEmailRegExp: RegExp = /^[a-z]+\.[1-9]\d*@osu\.edu$/;

/**
 * Regular expression for validating Buckeyemail addresses.
 * @type {RegExp}
 */
const buckeyemailRegExp: RegExp = /^[a-z]+\.[1-9]\d*@buckeyemail\.osu\.edu$/;

/**
 * Validates if the input string is in the OSU name.number format.
 * @param {string} nameDotNumber - The string to validate.
 * @returns {boolean} True if the string matches the OSU name.number format, otherwise false.
 */
export function isNameDotNumber(nameDotNumber: string): boolean {
  return osuNameDotNumberRegExp.test(nameDotNumber);
}

/**
 * Validates if the input string is an OSU email address.
 * @param {string} email - The email address to validate.
 * @returns {boolean} True if the email address matches the OSU email format, otherwise false.
 */
export function isOSUEmail(email: string): boolean {
  return osuEmailRegExp.test(email);
}

/**
 * Validates if the input string is a Buckeyemail address.
 * @param {string} buckeyemail - The email address to validate.
 * @returns {boolean} True if the email address matches the Buckeyemail format, otherwise false.
 */
export function isBuckeyemail(buckeyemail: string): boolean {
  return buckeyemailRegExp.test(buckeyemail);
}

/**
 * Validates if the input string is either an OSU email address or a Buckeyemail address.
 * @param {string} email - The email address to validate.
 * @returns {boolean} True if the email address matches either the OSU email format or the Buckeyemail format, otherwise false.
 */
export function isOSUOrBuckeyemail(email: string): boolean {
  return isOSUEmail(email) || isBuckeyemail(email);
}
