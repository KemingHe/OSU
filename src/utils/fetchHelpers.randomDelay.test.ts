// ./src/utils/fetchHelpers.randomDelay.test.ts
//
// Unittests for the randomDelay function in fetchHelpers.ts.

// Vitest essential imports.
import { describe, expect, it, vi } from "vitest";

// Testing target import.
import { type RandomDelayOptions, randomDelay } from "@src/utils/fetchHelpers";

// -----------------------------------------------------------------------------
// RandomDelayOption array for batch testing.
const validRandomDelayOptions: RandomDelayOptions[] = [
  { min: 0, max: 25 },
  { min: 25, max: 25 },
  { min: 25, max: 50 },
  { min: 50, max: 50 },
];

// Batch test helper function for randomDelay function.
async function batchRandomDelayExpectation(): Promise<void> {
  for (const options of validRandomDelayOptions) {
    await oneRandomDelayExpectation(options);
  }
}

async function oneRandomDelayExpectation({
  min,
  max,
}: RandomDelayOptions): Promise<void> {
  it(`returns a promise that resolves after a random delay within the specified range: {min: ${min}, max:${max}}`, async () => {
    // Monitor setTimeout calls.
    const setTimeoutSpy = vi.spyOn(globalThis, "setTimeout");

    // Call the randomDelay function.
    await randomDelay({ min, max });

    // Expect setTimeout to be called once
    // with a random delay within the specified range.
    expect(setTimeoutSpy).toHaveBeenCalledTimes(1);
    if (setTimeoutSpy.mock.calls?.[0]?.[1]) {
      const delay: number = setTimeoutSpy.mock.calls[0][1];
      expect(delay).toBeGreaterThanOrEqual(min);
      expect(delay).toBeLessThanOrEqual(max);
    } else {
      throw new Error("setTimeout not called with expected arguments.");
    }

    // Restore setTimeout spy.
    setTimeoutSpy.mockRestore();
  });
}

// Test suite for randomDelay function.
describe("The randomDelay fetch helper function", () => {
  batchRandomDelayExpectation();

  it("throws an error if minimum delay is less than 0", async () => {
    await expect(
      randomDelay({ min: -1, max: 0 }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("throws an error if maximum delay is less than or equal to 0", async () => {
    await expect(
      randomDelay({ min: 0, max: -1 }),
    ).rejects.toThrowErrorMatchingSnapshot();
    await expect(
      randomDelay({ min: 0, max: 0 }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("throws an error if maximum delay is less than minimum delay", async () => {
    await expect(
      randomDelay({ min: 100, max: 50 }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
