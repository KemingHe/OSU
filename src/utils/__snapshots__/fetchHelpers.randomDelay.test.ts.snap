// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The randomDelay fetch helper function > throws an error if maximum delay is less than minimum delay 1`] = `[Error: Maximum delay must be greater than or equal to minimum delay.]`;

exports[`The randomDelay fetch helper function > throws an error if maximum delay is less than or equal to 0 1`] = `[Error: Maximum delay must be strictly greater than 0.]`;

exports[`The randomDelay fetch helper function > throws an error if maximum delay is less than or equal to 0 2`] = `[Error: Maximum delay must be strictly greater than 0.]`;

exports[`The randomDelay fetch helper function > throws an error if minimum delay is less than 0 1`] = `[Error: Minimum delay must be greater than or equal to 0.]`;
